xbmc/games/dialogs/osd/DialogInGameSaves.cpp

```diff
@@ -0,0 +1,104 @@
+/*
+ *  Copyright (C) 2020 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "DialogInGameSaves.h"
+#include "cores/RetroPlayer/playback/IPlayback.h"
+#include "cores/RetroPlayer/savestates/ISavestate.h"
+#include "cores/RetroPlayer/savestates/SavestateDatabase.h"
+#include "cores/RetroPlayer/guibridge/GUIGameRenderManager.h"
+#include "cores/RetroPlayer/guibridge/GUIGameSettingsHandle.h"
+#include "guilib/LocalizeStrings.h"
+#include "guilib/WindowIDs.h"
+#include "ServiceBroker.h"
+#include "settings/GameSettings.h"
+#include "settings/MediaSettings.h"
+#include "utils/log.h"
+//#include "utils/Variant.h"
+
+using namespace KODI;
+using namespace GAME;
+using namespace RETRO;
+
+CDialogInGameSaves::CDialogInGameSaves()
+    : CDialogGameVideoSelect(WINDOW_DIALOG_IN_GAME_SAVES)
+{
+}
+
+std::string CDialogInGameSaves::GetHeading()
+{
+  return g_localizeStrings.Get(35249); // "Saved games"
+}
+
+void CDialogInGameSaves::PreInit()
+{
+  m_items.Clear();
+
+  InitSavedGames();
+
+  CFileItemPtr item = std::make_shared<CFileItem>(g_localizeStrings.Get(223)); // "New"
+  item->SetIconImage("DefaultAddSource.png");
+  item->SetPath("");
+
+  m_items.AddFront(std::move(item), 0);
+}
+
+void CDialogInGameSaves::InitSavedGames()
+{
+  auto gameSettings = CServiceBroker::GetGameRenderManager().RegisterGameSettingsDialog();
+
+  // save current game
+  IPlayback* playback = gameSettings->GetPlayback();
+  playback->CreateSavestate();
+
+  CSavestateDatabase db;
+  db.GetSavestatesNav(m_items, gameSettings->GetPlayingGame(), gameSettings->GameClientID());
+
+  m_items.Sort(SortByDate, SortOrderDescending);
+}
+
+void CDialogInGameSaves::GetItems(CFileItemList& items)
+{
+  for (const auto& item : m_items)
+    items.Add(item);
+}
+
+void CDialogInGameSaves::OnItemFocus(unsigned int index)
+{
+  if (static_cast<int>(index) < m_items.Size())
+    m_focusedItemIndex = index;
+}
+
+unsigned int CDialogInGameSaves::GetFocusedItem() const
+{
+  return m_focusedControl;
+}
+
+void CDialogInGameSaves::PostExit()
+{
+  m_items.Clear();
+}
+
+void CDialogInGameSaves::OnClickAction()
+{
+  if (static_cast<int>(m_focusedItemIndex) < m_items.Size())
+  {
+    auto gameSettings = CServiceBroker::GetGameRenderManager().RegisterGameSettingsDialog();
+    IPlayback* playback = gameSettings->GetPlayback();
+    std::string savePath = m_items[m_focusedItemIndex]->GetPath(); 
+
+    if (savePath.empty())
+    {
+      playback->ResetLoadedSave();
+      gameSettings->ResetGame();
```

Comment:
I hit this logic path when I click "New". Shouldn't this be `CreateSavestate()`?

Reply:
`ResetLoadedSave()` resets the loaded game path to `""` so when RetroPlayer calls `CreateSavestate()` it will save to a new path

Reply:
I see. The "New" logic is a bit unintuitive for me but we'll see what testers think. One of these two commands is calling a hardware reset, which we probably don't want (I don't expect creating a new savestate to blow out the current state)

Reply:
The current state is saved when you enter the in-game manager [here](https://github.com/garbear/xbmc/blob/b668f5352125b81ffff4e80535c99462434ff2af/xbmc/games/dialogs/osd/DialogInGameSaves.cpp#L54-L56) and then the hardware reset is what is called when you reset the game from the osd, I don't know how else to restart the game